pandas/tools/merge.py

```diff
@@ -112,44 +183,152 @@ def ordered_merge(left, right, on=None, left_by=None, right_by=None,
-        for key, xpiece in left.groupby(left_by):
-            merged = _merger(xpiece, right)
-            for k in left_by:
-                # May have passed ndarray
-                try:
-                    if k in merged:
-                        merged[k] = key
-                except:
-                    pass
-            pieces.append(merged)
-        return concat(pieces, ignore_index=True)
+        result, _ = _groupby_and_merge(left_by, left, right,
+                                       lambda x, y: _merger(x, y))
     elif right_by is not None:
-        if not isinstance(right_by, (list, tuple)):
-            right_by = [right_by]
-        pieces = []
-        for key, ypiece in right.groupby(right_by):
-            merged = _merger(left, ypiece)
-            for k in right_by:
-                try:
-                    if k in merged:
-                        merged[k] = key
-                except:
-                    pass
-            pieces.append(merged)
-        return concat(pieces, ignore_index=True)
+        result, _ = _groupby_and_merge(right_by, right, left,
+                                       lambda x, y: _merger(y, x))
     else:
-        return _merger(left, right)
+        result = _merger(left, right)
+    return result
+
+
+def asof_merge(left, right, on=None,
+               left_on=None, right_on=None,
+               by=None,
+               suffixes=('_x', '_y'),
+               tolerance=None, strict=False):
+    """Perform an asof merge. This by definition will fill the right data with
+    the asof the left data. The data must be ordered according to the 'on'
+    (typically time). Optionally perform group-wise merge.
+
+    .. versionadded 0.18.2
+
+    Parameters
+    ----------
+    left: DataFrame
+    right: DataFrame
+        Data will be de-duplicated on the [by, on] combination if needed.
+    on: label or list
+        Field names to join on. Must be found in both DataFrames.
+        The data MUST be ordered. Furthermore this must be a datetimelike or
+        integer-index.
+    left_on: label or list, or array-like
+        Field names to join on in left DataFrame. Can be a vector or list of
+        vectors of the length of the DataFrame to use a particular vector as
+        the join key instead of columns
+    right_on: label or list, or array-like
+        Field names to join on in right DataFrame or vector/list of vectors per
+        left_on docs
+    by: column name or list of column names
+        Group both the left and right DataFrames by the group columns; perform
+        the merge operation on these pieces and recombine.
+    suffixes: 2-length sequence (tuple, list, ...)
+        Suffix to apply to overlapping column names in the left and right
+        side, respectively
+    tolerance: integer or Timedelta, optional, default None
+        select asof tolerance within this range, must be compatible
+        to the merge index.
+    strict: boolean, default False
```

Comment:
how about one of these:

1) ``strict='less_than_or_equal' | 'less_than'``
2) ``match='less_than_or_equal' | 'less_than'``
3) ``match_less_that_or_equal=True | False``
4) ``allow_exact_matches=True``